### ISSUE TITLE
Enhanced authentication

### ISSUE THREAD

**Issue by robertsLando** [moscajs/aedes#833]

**Comment by BenjaminDobler** [comment]
@robertsLando — as agreed on the #821 plan, here's the proposed hook API for Enhanced Authentication before I write code.

Today a CONNECT carrying an `Authentication Method` is rejected with `0x8C` (`connect.js`, §4.12 guard). This replaces that with an opt-in broker hook that drives the `AUTH`-packet challenge/response exchange. I've kept it a **single callback, called once per exchange step**, matching aedes's existing `authenticate` / `authorize*` style — the app keeps its own per-exchange state (keyed by `client`, e.g. a SCRAM/OAuth session), the same way plugin-based brokers do.

## Handler

```js
aedes.authenticateEnhanced = function (client, method, data, callback) {
  // method: <string>            Authentication Method from CONNECT (constant for the exchange)
  // data:   <Buffer>|undefined  Authentication Data for THIS step — from CONNECT on the
  //                             first call, then from each client AUTH packet
  //
  // callback(error, result):
  //   error (truthy)            -> auth failed / reject. error.reasonCode (default 0x87
  //                                Not authorized; e.g. 0x8C Bad authentication method),
  //                                error.reasonString optional.
  //   { done: false, data }     -> CONTINUE: server sends AUTH (0x18 Continue) + method +
  //                                data, then awaits the client's next AUTH.
  //   { done: true, data? }     -> SUCCESS: accept + optional final Authentication Data.
}
```

## Flow

- **Initial (during CONNECT)** — CONNECT has an `Authentication Method` ⇒ broker calls `authenticateEnhanced(client, method, connectAuthData, cb)`.
  - `{ done: false, data }` ⇒ server sends `AUTH 0x18` + `Authentication Method` + `data`; client replies with `AUTH 0x18` + data ⇒ handler is called again with the new data. Repeat.
  - `{ done: true, data? }` ⇒ `CONNACK 0x00` (+ optional `Authentication Data`); the connection proceeds normally.
  - `error` ⇒ `CONNACK` with `error.reasonCode` / `reasonString`.
- **Re-authentication (§4.12.1)** — a connected client sends `AUTH 0x19` (Re-authenticate) ⇒ the same handler drives a fresh exchange; success ⇒ server sends `AUTH 0x00` (Success); `error` ⇒ server `DISCONNECT` with the reason.
- **No handler configured** — unchanged: a CONNECT with an `Authentication Method` is still rejected with `0x8C` (back-compat). The username/password `authenticate` hook is used only when there is **no** Authentication Method.

## Plumbing (once the API is agreed)

- Route inbound `AUTH` packets in `handlers/index.js` ⇒ new `handlers/auth.js`.
- Gate the existing `0x8C` guard on "no `authenticateEnhanced` configured".
- Tests: two-step SCRAM-style exchange, immediate success, failure ⇒ CONNACK reason, and (if in scope) re-auth success/fail.

## Open questions for you

1. **Separate handler vs. folding into `authenticate`?** I lean **separate** (`authenticateEnhanced`) — the shapes differ (method-based, multi-step, binary data), and it keeps the common username/password path untouched. OK, or would you rather one unified handler?
2. **Result shape:** `{ done, data }` + error-for-failure — or do you prefer an explicit `{ action: 'continue' | 'success' | 'failure', ... }`?
3. **Re-auth scope:** in the first PR, or land initial-auth first and do re-auth (`AUTH 0x19` inbound + DISCONNECT-on-fail) as a fast follow-up? Initial auth alone already covers the common SASL-at-connect case.

Happy to adjust any of it before I start. 👍

**Comment by robertsLando** [comment]
Thanks, this is a solid write-up and I'm on board with the overall direction. The API shape matches aedes's plugin style well — a single per-step callback with app-owned state keyed by `client`, failure via the error object — and the wire layer is already there (`mqtt-packet@9.0.2` encodes/decodes AUTH with reason codes `0x00/0x18/0x19` and the `authenticationMethod`/`authenticationData` properties).

My main concern isn't the API, it's the **plumbing** — that section is thin, and it's where the real difficulty lives. A few things I want pinned down before you start writing code, because a couple of them can force the handler's shape.

### Blocking concerns

**1. Inbound-AUTH-during-CONNECT dispatch — watch for a handshake deadlock.**
Continuation AUTH packets arrive *after* CONNECT but *before* CONNACK. In that window `connackSent === false` and `connected === false`, and our parser's `enqueue` (`lib/client.js`) routes packets to a queue that only drains on the `'connected'` event — which won't fire until the exchange completes. So the client's next AUTH can sit queued waiting for a `'connected'` that is itself waiting for that AUTH. The pre-switch gate in `handlers/index.js` (`if (!client.connecting && !client.connected)`) and the fastq ordering all assume nothing but the connect pipeline runs pre-CONNACK. Please make the plan state explicitly how AUTH is dispatched in the pre-CONNACK window (e.g. special-case `cmd === 'auth'` in `enqueue`). This is the first thing I'd prototype and verify end-to-end.

**2. How does the exchange suspend the connect pipeline?**
`authenticate` today is one action in the `runSeries` chain in `connect.js` that `pause()`s and resolves via a single callback. Enhanced auth turns that one step into an N-round-trip exchange handled in a *different* file (`auth.js`). The plan should say how `auth.js` resumes the connect continuation and eventually reaches `completeConnect`/`doConnack` — e.g. the connect action stashes the `runSeries` `done` on the client and `auth.js` invokes it on `{ done: true }`, reusing the existing `pause()`/`resume()`.

### Spec conformance (please fold into the plan)

**3. aedes must enforce that the Authentication Method is constant across the exchange (§4.12).** Good call hiding `method` from the app on continuation calls — but that means aedes itself has to compare each inbound AUTH's method against the stored one and reject a mismatch (protocol error). Not optional.

**4. Re-auth eligibility.** A client that authenticated with username/password must not be allowed to start a re-auth (`0x19`) exchange — only clients whose CONNECT carried an Authentication Method may. Needs the method stored at connect time and a protocol-error DISCONNECT for ineligible `0x19`. Worth a line even though re-auth is deferred.

**5. CONNACK has no reason-string path today.** Reason strings are currently only emitted on PUBACK/PUBREC/SUBACK/DISCONNECT — the `Connack` class sets none. So `error.reasonString` on an *initial-auth* failure needs new plumbing to land on the CONNACK. The success side is nearly free: `{ done: true, data }` → CONNACK `authenticationData` works because `Connack` already forwards `arg.properties`.

**6. Timeout / DoS.** Please confirm the existing connect timeout still fires across the multi-round window so a client that opens CONNECT-with-method and never finishes can't leave a half-open connection pinned.

### Answers to your open questions

1. **Separate handler — yes.** Method-based + multi-step + binary genuinely doesn't fold into `authenticate(client, username, password)`. Just note in the PR that this is the first v5 feature to add a new hook (Server Reference and Reason Strings reused the existing `authenticate`/`preConnect` error) — the justification is fine, worth stating. One nit: our hooks are verb+noun (`authorizePublish`, `authorizeSubscribe`); `authenticateEnhanced` reads a little off but I won't block on the name.
2. **`{ done, data }` + error-for-failure — keep it.** More consistent with the codebase than an action enum. Only thing I'd consider is an optional `properties` field so the app can attach User Properties / reason string to an outbound *continue* AUTH — spec-legal, rarely needed, and adding it later would be breaking. Your call whether to include now or defer.
3. **Land initial-auth first, re-auth as fast-follow — agree.** Just design the client state (`_authenticationMethod`) and the `auth.js` shape now so re-auth slots in without rework.

### Housekeeping

- Add `0x18`/`0x19` to `lib/constants.js` (they only live in mqtt-packet today).
- The two negative tests that currently lock in the rejection (`test/mqtt5.js`, the "rejected with 0x8C" and the auth-data-without-method cases) need to be **revised**, not just added alongside.
- Standard v5 recipe still applies: new option on `aedes.js` + `types/instance.d.ts` + `test/types/aedes.test-d.ts` + `docs/Aedes.md`.

Happy with the API — just want the pre-CONNACK dispatch and the pipeline-suspension mechanism spelled out before code. 👍

**Comment by BenjaminDobler** [comment]
Thanks — this is exactly the level I wanted to nail before coding. Confirmed each concern against the machinery; here's the revised plan with the plumbing spelled out.

## 1. Pre-CONNACK AUTH dispatch (the deadlock) — confirmed & solved

You're right. `enqueue` (`lib/client.js`) dispatches only packet #1 (`_parsingBatch === 1`, the CONNECT) or anything after `connackSent`; everything in between is pushed to `_parser._queue`, which drains only on `'connected'` → `dequeue`. A continuation AUTH lands in that gap and waits for a `'connected'` that is waiting for it. Deadlock.

**Fix:** special-case AUTH in the pre-CONNACK window, in `enqueue`:

```js
if (client.connackSent || client._parsingBatch === 1 ||
    (packet.cmd === 'auth' && client.connecting)) {
  handle(client, packet, client._nextBatch)
} else { /* queue as today */ }
```

Then in `handlers/index.js`: AUTH already clears the `!connecting && !connected` gate (mid-exchange `connecting === true`), so I add `case 'auth': handleAuth(...)` to the switch. Non-AUTH packets sent mid-exchange keep today's behavior (queued until `'connected'`), so nothing else runs pre-CONNACK.

## 2. Suspending / resuming the connect pipeline

The `authenticate` action `pause()`s the read loop and resolves the `runSeries` via one `done`. Enhanced auth turns that into N round-trips owned by `auth.js`, so:

- The enhanced-auth connect action stashes the `runSeries` continuation on the client (`client._enhancedAuth = { method, done }`) and, on `{ done: false, data }`, writes `AUTH 0x18` + method + data and **`resume()`s the read loop** (the `authenticate` pause would otherwise leave the client's reply unread in the socket buffer).
- Each continuation AUTH → `handleAuth` → `authenticateEnhanced` again. `{ done: false }` → another `AUTH 0x18` + `resume()`. `{ done: true }` → `handleAuth` invokes the stashed `done()` → the pipeline resumes at `setKeepAlive` → … → `doConnack` (CONNACK `0x00`). `error` → stashed `done(err)` → rejection CONNACK.

I'll **prototype this end-to-end first** (a 2-round SCRAM-style exchange) and verify the pause/resume + dispatch interaction before building the rest — agreed it's the risky part.

## 3. Authentication Method constant across the exchange [§4.12] — enforced by aedes

Stored as `client._authenticationMethod` at connect. `handleAuth` compares each inbound AUTH's `authenticationMethod` to it and, on mismatch (or an AUTH with no method), rejects with a `0x82` Protocol Error — DISCONNECT if connected (re-auth), CONNACK during initial auth. The app never sees the method on continuation calls (it's constant), so aedes owns the check.

## 4. Re-auth eligibility (deferred, but designed now)

`client._authenticationMethod` is set only when CONNECT carried an Authentication Method. An `AUTH 0x19` from a client without it → `0x82` Protocol Error DISCONNECT. The `auth.js` shape and `_enhancedAuth`/`_authenticationMethod` state are designed so re-auth slots in as the fast-follow with no rework (it reuses `handleAuth`, terminating with `AUTH 0x00` on success / DISCONNECT on failure instead of a CONNACK).

## 5. CONNACK reason string

`Connack` already forwards `arg.properties`, so the success side (`{ done: true, data }` → CONNACK `authenticationData`) is free. For an initial-auth **failure**, I'll pass `properties: { reasonString }` through the existing `rejectConnect` path (it already carries `properties`, from #838). CONNACK/DISCONNECT are exempt from Request Problem Information, so no gating. (This is genuinely new — no CONNACK sets a reason string today — but it's wiring, not a new mechanism.)

## 6. Timeout / DoS — confirmed a gap, will re-arm

`handleConnect` clears `client._connectTimer` on the first byte of CONNECT, so today nothing bounds a stalled multi-round exchange. I'll arm an auth-exchange timer (bounded by `connectTimeout`) when the first `{ done: false }` is returned, reset it on each round, and clear it on completion; on expiry → close (DISCONNECT `0x87` for a connected re-auth). A never-finished CONNECT-with-method can't pin a half-open connection.

## Your open questions

1. **Separate handler — keeping `authenticateEnhanced`.** I'll note in the PR that it's the first v5 feature to add a new hook (Server Reference / Reason Strings reused `preConnect`/`authenticate` errors). Fair on the verb+noun nit — alternatives are `enhancedAuthenticate` / `authenticateV5`; I'll go with `authenticateEnhanced` unless you'd prefer one of those.
2. **`{ done, data }` + error-for-failure — keeping it, and I'll include the optional `properties` now.** Since adding it later would be breaking, the result becomes `{ done, data?, properties? }` — `properties` lets the app attach User Properties / a reason string to an outbound continue-AUTH. Cheap to include, so I'd rather not leave a breaking gap.
3. **Initial-auth first, re-auth fast-follow — agreed**, with the state/`auth.js` shape designed for it now (per #4).

## Housekeeping (folded in)

- Add `0x18`/`0x19` to `lib/constants.js`.
- **Revise** the two negative tests (`… rejected with 0x8C`, auth-data-without-method) — they only hold when `authenticateEnhanced` is unset, so they become "no-handler ⇒ still rejected" cases.
- Standard v5 recipe: `aedes.js` option + `types/instance.d.ts` + `test/types/aedes.test-d.ts` + `docs/Aedes.md`.

Plan: prototype the §1/§2 dispatch + suspension end-to-end, confirm here, then the full PR (constants, `auth.js`, `handleAuth`, connect-action + timer, gated `0x8C`, tests, types, docs). Sound good?

**Comment by robertsLando** [comment]
This all checks out against the code — green light on the prototype. I traced each piece:

- **§1 deadlock** — confirmed. `enqueue` only dispatches `_parsingBatch === 1` or `connackSent`; everything else queues in `_parser._queue`, drained by `dequeue` on `'connected'`, which only fires after `doConnack`. Your `packet.cmd === 'auth' && client.connecting` guard is valid — `handleConnect` sets `connecting = true` before the pipeline. And AUTH already clears the `!connecting && !connected` gate in `handlers/index.js` but falls through to `default → destroy`, so `case 'auth'` is exactly the gap.
- **§2 pause/resume** — confirmed. `authenticate` pauses; while `_paused`, `nextBatch` bails so the socket isn't drained, so the `resume()` is genuinely required, and `resume()` calls `_nextBatch()` to drain buffered bytes. Stashing the `runSeries` continuation across N round-trips is the right extension — just keep it distinct from the per-packet `_nextBatch` done (looks like you already do).
- **§6 timer** — confirmed, `handleConnect` clears `_connectTimer` up front, so nothing bounds a stalled exchange today. Re-arming bounded by `connectTimeout` is right.
- **§3/§4** spec-accurate, aedes owning the method check is the right call. Constants `0x18`/`0x19` are indeed missing today.

**One correction on §5 — the success `authenticationData` is *not* free.** `doConnack` only builds the v5 capability properties when `!arg.properties`. If you set `arg.properties = { authenticationData }` on the success path, you skip that block and drop `sharedSubscriptionAvailable: false` ([MQTT-3.2.2-13]), the assigned client id, Server Keep Alive, the clamped Session Expiry, and Response Information (#1091). So **merge** `authenticationData` into the generated properties inside `doConnack` (e.g. stash it on the client and append after `connackProperties()`) rather than passing `arg.properties` through. The failure path via `rejectConnect(..., properties: { reasonString })` is fine as-is.

Minor: the "auth-data-without-method → `0x82`" reject holds regardless of enhanced-auth support (it's always a protocol error), so leave that test unconditional — only the `0x8C` "method present, no handler" case gets gated on `authenticateEnhanced`.

Otherwise 👍 — prototype the §1/§2 dispatch + suspension end-to-end, confirm here, then go for the full PR. `authenticateEnhanced` is fine as the name; `{ done, data?, properties? }` with the optional `properties` up front to avoid a breaking gap is the right call.

**Comment by BenjaminDobler** [comment]
Prototype done and **§1/§2 verified end-to-end** — a two-round `AUTH` exchange completes a CONNACT. 368 unit tests + lint green.

### What the prototype does (matches the plan, with your §5 fix)

- **§1 dispatch** — `enqueue` now dispatches immediately when `packet.cmd === 'auth' && client.connecting`; `handlers/index.js` routes `case 'auth' → handleAuth`. The continuation AUTH is no longer queued behind a `'connected'` that can't fire.
- **§2 suspension** — the `authenticate` step calls `startEnhancedAuth` and does **not** resolve; `auth.js` `runEnhancedAuthStep` `pause()`s while the hook is pending, and on `{ done: false }` writes `AUTH 0x18` + `resume()`s so the client's reply is read and dispatched. On `{ done: true }` it invokes the stashed connect continuation → the pipeline runs on to `doConnack`. The `runSeries` continuation is stashed on `client._enhancedAuth` and is distinct from the per-packet `_nextBatch`, as you flagged. No circular import: `connect.js` hands `auth.js` `onSuccess`/`onFailure` closures rather than `auth.js` reaching into `doConnack`.
- **§5 fix applied** — success stashes `client._authenticationData`, and `doConnack` **merges** it into the generated properties inside the capability block. Test asserts the CONNACK carries both `authenticationData: 'server-final'` **and** `sharedSubscriptionAvailable: false`.

### The verifying test

Raw v5 client: `CONNECT` (method `SCRAM-SHA-256`, data `client-first`) → server `AUTH 0x18` (`server-challenge`) → client `AUTH 0x18` (`client-final`) → `CONNACK 0x00` (`server-final` + capabilities). The hook is called exactly twice.

### Still to do for the full PR (not in the prototype)

- **§6 timer** — arm an auth-exchange timeout bounded by `connectTimeout` (currently a stalled exchange is unbounded).
- **Error paths** — reject → CONNACK reason code + `reasonString` via `rejectConnect`; §3 method-mismatch → `0x82`; a stray/oversized AUTH.
- **§4 re-auth** — `AUTH 0x19` inbound + `AUTH 0x00`/DISCONNECT terminals + eligibility check (deferred, but the `auth.js`/`_authenticationMethod` shape already supports it).
- **Result `properties`** — the optional field is threaded to the outbound continue-AUTH; will add a test.
- **Housekeeping** — the two negative tests (leave "auth-data-without-method → 0x82" unconditional per your note; gate only "method present, no handler → 0x8C"), plus `aedes.js` option docs / `types/instance.d.ts` / `test-d` / `docs/Aedes.md`.

Prototype is on `feat/v5-enhanced-auth` (`7eb9d61`). Going ahead with the full PR on top unless you want to eyeball the dispatch/suspension diff first. 👍

**Comment by BenjaminDobler** [comment]
Full PR is up: #1119 (against `mqttv5`).

On top of the validated prototype it adds:

- **§6 timer** — each round bounded by `connectTimeout`; a stalled exchange (client never answers a challenge) is closed.
- **Error paths** — rejecting hook → CONNACK `reasonCode` (default 0x87) + `reasonString`; mid-exchange Authentication Method change → 0x82; a synchronously-throwing hook is contained as an auth failure.
- **Negative tests revised per your note** — 0x8C is gated on the missing `authenticateEnhanced` handler; auth-data-without-method stays an unconditional 0x82.
- **Result `properties`** threaded to the challenge `AUTH` (+ test).
- **Types + docs** — `authenticateEnhanced` handler type, `test-d`, and a `docs/Aedes.md` section.

**Re-auth** (client-initiated `AUTH 0x19` after connect) is left as the agreed fast-follow — a stray/re-auth `AUTH` is cleanly rejected with a 0x82 DISCONNECT for now, and `_authenticationMethod` state already supports the follow-up.

Full suite + lint + tsd green.
